CMakeLists.txt

```diff
@@ -1,316 +1,119 @@
+  DESCRIPTION "C++ database access library"
+  HOMEPAGE_URL "https://soci.sourceforge.net/"
+  LANGUAGES CXX
+)
 
-    if (NOT supported)
-        message(STATUS "IPO / LTO not supported")
-    endif()
+include(soci_utils)
 
-    if (supported AND NOT SOCI_ASAN)
-        message(STATUS "IPO / LTO enabled")
-        set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
-
-        if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-            # Check for lld as clang lto works best with its own linker
-            soci_use_ld_if_supported(lld)
-        endif()
-    elseif(supported)
-        message(STATUS "IPO / LTO is supported but conflicts with ASAN and not enabled")
-    endif()
+if (NOT DEFINED CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 14)
+  set(CMAKE_CXX_STANDARD 14)
 endif()
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 
-if (SOCI_VISIBILITY)
-    # Test whether visibility is supported
-    include(CheckCSourceCompiles)
-    check_c_source_compiles(
-        "
-            __attribute__ (( visibility(\"default\") )) int f1() { return 0; }
-            __attribute__ (( visibility(\"hidden\") ))  int f2() { return 1; }
+include(CMakeDependentOption)
+include(CheckIPOSupported)
 
-            int main(int argc, char* argv[]) { f1(); f2(); return 0; }
-        "
-        SOCI_HAVE_VISIBILITY_SUPPORT
-    )
+check_ipo_supported(RESULT LTO_AVAILABLE)
 
-    if (SOCI_HAVE_VISIBILITY_SUPPORT)
-        message(STATUS "gcc / clang visibility enabled")
-        set(CMAKE_CXX_VISIBILITY_PRESET hidden)
-        cmake_policy(SET CMP0063 NEW)
-    endif()
+if (SOCI_STATIC)
+  set(SHARED_DEFAULT OFF)
+elseif(DEFINED BUILD_SHARED_LIBS)
+  set(SHARED_DEFAULT ${BUILD_SHARED_LIBS})
 else()
-    set(SOCI_HAVE_VISIBILITY_SUPPORT off)
+  set(SHARED_DEFAULT ON)
 endif()
 
-# Allow using alternative linker such as mold, which can be significantly
-# faster than GNU ld.
-option(SOCI_LD "Use non-default linker, such as 'mold'" "")
-if(SOCI_LD)
-	# -fuse-ld works only with recent gcc (>= 12), but we don't need to support
-	# this with all gcc versions as this is entirely optional. I.e. if we really
-	# wanted to, we could use -B option as explained in mold README to make it
-	# work with any gcc, but for now keep the things simple.
-	soci_use_ld_if_supported(${SOCI_LD})
-endif()
-
-###############################################################################
-# SOCI configuration
-###############################################################################
-include(SociConfig)
-
-colormsg(_HIBLUE_ "Configuring SOCI:")
+option(SOCI_SHARED "Enable building SOCI as a shared library" ${SHARED_DEFAULT})
+option(SOCI_TESTS "Enable building SOCI test cases" ${PROJECT_IS_TOP_LEVEL})
+cmake_dependent_option(SOCI_LTO "Enable link time optimizations in release builds" ON "LTO_AVAILABLE" OFF)
+option(SOCI_VISIBILITY "Make all functions hidden by default - this exposes only explicitly exported functions" ON)
```

Comment:
This is not very important, but I think we could drop this option and just always use hidden visibility if it's supported because I don't see any reason to ever disable this.

Reply:
Fine by me :+1: